### PR TITLE
New DSA signing API for #1648

### PR DESCRIPTION
**Pull request by joernheissler**

### CODE REVIEW

File: docs/hazmat/primitives/asymmetric/dsa.rst [hunk 1]
```diff
@@ -74,8 +74,55 @@ provider.
...
+    to the size of DSA's "q" parameter, which is just a few bytes.
+
+.. doctest::
+
+    >>> digester = hashes.Hash(hashes.SHA256(), backend=default_backend())
+    >>> digester.update(data)
+    >>> digest = digester.finalize()
+    >>> signature = private_key.sign(digest, already_hashed=True)
+
+Specifying the per-message key k:
```
Comment (reaperhulk) [inline docs/hazmat/primitives/asymmetric/dsa.rst]: Under what scenario do you want the ability to set `k` manually? I realize there's a warning below, but this API is incredibly dangerous.

Comment (alex) [inline docs/hazmat/primitives/asymmetric/dsa.rst]: I would say setting k is out of scope for this API concern (and separately,
we should never support it)

On Wed, Jan 6, 2016 at 12:49 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/hazmat/primitives/asymmetric/dsa.rst
> <https://github.com/pyca/cryptography/pull/2642#discussion_r48986337>:
>
> > +
> > +Signing a precomputed digest:
> > +
> > +.. warning::
> > +
> > +    Only use this for digests, do NOT sign raw data. The data will get truncated
> > +    to the size of DSA's "q" parameter, which is just a few bytes.
> > +
> > +.. doctest::
> > +
> > +    >>> digester = hashes.Hash(hashes.SHA256(), backend=default_backend())
> > +    >>> digester.update(data)
> > +    >>> digest = digester.finalize()
> > +    >>> signature = private_key.sign(digest, already_hashed=True)
> > +
> > +Specifying the per-message key k:
>
> Under what scenario do you want the ability to set k manually? I realize
> there's a warning below, but this API is incredibly dangerous.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2642/files#r48986337>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (joernheissler) [inline docs/hazmat/primitives/asymmetric/dsa.rst]: For debugging and educational purposes. And perhaps to allow users to implement stuff like rfc6979. Openssl implemented something similar which uses the private key + message digest do seed a PRNG. But it's non-deterministic. I'm not saying that this approach is bad, but it's not rfc6979.

What's wrong with supporting a dangerous API? For one, it's all in "hazmat". And with the big fat warning even a stupid user shouldn't accidentally use this function.
Perhaps change "warning" to "dangerous"?

Comment (alex) [inline docs/hazmat/primitives/asymmetric/dsa.rst]: There's no reason for it. pyca/cryptography, we shouldn't be giving people
more tools to shoot themselves in the foot than is _absolutely_ necessary.

On Wed, Jan 6, 2016 at 1:33 PM, joernheissler <notifications@github.com>
wrote:

> In docs/hazmat/primitives/asymmetric/dsa.rst
> <https://github.com/pyca/cryptography/pull/2642#discussion_r48991323>:
>
> > +
> > +Signing a precomputed digest:
> > +
> > +.. warning::
> > +
> > +    Only use this for digests, do NOT sign raw data. The data will get truncated
> > +    to the size of DSA's "q" parameter, which is just a few bytes.
> > +
> > +.. doctest::
> > +
> > +    >>> digester = hashes.Hash(hashes.SHA256(), backend=default_backend())
> > +    >>> digester.update(data)
> > +    >>> digest = digester.finalize()
> > +    >>> signature = private_key.sign(digest, already_hashed=True)
> > +
> > +Specifying the per-message key k:
>
> For debugging and educational purposes. And perhaps to allow users to
> implement stuff like rfc6979. Openssl implemented something similar which
> uses the private key + message digest do seed a PRNG. But it's
> non-deterministic. I'm not saying that this approach is bad, but it's not
> rfc6979.
>
> What's wrong with supporting a dangerous API? For one, it's all in
> "hazmat". And with the big fat warning even a stupid user shouldn't
> accidentally use this function.
> Perhaps change "warning" to "dangerous"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2642/files#r48991323>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: docs/hazmat/primitives/asymmetric/dsa.rst [hunk 2]
```diff
@@ -74,8 +74,55 @@ provider.
...
+
+    >>> digester = hashes.Hash(hashes.SHA256(), backend=default_backend())
+    >>> digester.update(data)
+    >>> signature = private_key.sign(digester)
+
+One-line signing:
+
+.. doctest::
+
+    >>> signature = private_key.sign(data, hashes.SHA256())
```
Comment (reaperhulk) [inline docs/hazmat/primitives/asymmetric/dsa.rst]: This (and the HashContext and precomptued digest forms) need to take padding as an arg still. We could allow pure raw signing with padding=None, but without information about the padding we can't construct the bytestring that will actually be signed.

Comment (joernheissler) [inline docs/hazmat/primitives/asymmetric/dsa.rst]: I don't think that DSA uses any padding.

Comment (reaperhulk) [inline docs/hazmat/primitives/asymmetric/dsa.rst]: Derp, sorry I was expanding this to RSA in my mind. You are correct that there's no padding arg for DSA (or ECDSA).

**Comment by reaperhulk** [comment]
This has been partially addressed by #2945 (in terms of API design, although that only implements RSA and does not allow prehashing at this time). I'm going to go ahead and close this PR and if we want to pursue this further we can discuss it on a new and more narrowly focused PR.

**Pull request closed without merging**